controllers/hostingdeploymentautoscalingjob/hostingdeploymentautoscalingjob_controller.go

```diff
@@ -0,0 +1,455 @@
+	if ctx.ScalableTargetDescriptionList, ctx.ScalingPolicyDescriptionList, err = r.describeAutoscalingPolicy(ctx); err != nil {
+		return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to describe HostingDeploymentAutoscaling."))
+	}
+
+	var action controllers.ReconcileAction
+	if action, err = r.determineActionForAutoscaling(ctx); err != nil {
+		return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to determine action for HostingDeploymentAutoscaling."))
+	}
+	ctx.Log.Info("Determined action for AutoscalingJob", "action", action)
+
+	// If update or delete, delete the existing Policy.
+	if action == controllers.NeedsDelete || action == controllers.NeedsUpdate {
+		if err = r.deleteAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to deleteAutoscalingPolicy"))
+		}
+
+		// Delete succeeded, set Description to nil.
+		ctx.ScalingPolicyDescriptionList = nil
+		ctx.ScalableTargetDescriptionList = nil
+	}
+
+	// If update or create, create the desired hostingdeploymentautoscaling.
+	if action == controllers.NeedsCreate || action == controllers.NeedsUpdate {
+		if ctx.ScalableTargetDescriptionList, ctx.ScalingPolicyDescriptionList, err = r.applyAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to applyAutoscalingPolicy"))
+		}
+	}
+
+	// Update the status accordingly.
+	// Review: Deleted Status may be unnecessary
+	status := CreatedAutoscalingJobStatus
+	if ctx.ScalingPolicyDescriptionList == nil {
+		status = DeletedAutoscalingJobStatus
+	}
+	if err = r.updateStatus(ctx, status); err != nil {
+		return err
+	}
+
+	// Remove the Finalizer on delete
+	if controllers.HasDeletionTimestamp(ctx.HostingDeploymentAutoscalingJob.ObjectMeta) {
+		if err = r.removeFinalizer(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Removes the finalizer held by our controller.
+func (r *Reconciler) removeFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	ctx.HostingDeploymentAutoscalingJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.HostingDeploymentAutoscalingJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+	if err = r.Update(ctx, ctx.HostingDeploymentAutoscalingJob); err != nil {
+		return errors.Wrap(err, "Failed to remove finalizer")
+	}
+	ctx.Log.Info("Finalizer has been removed")
+
+	return err
+}
+
+// getResourceIDListfromInputSpec converts the list of resources into a string list to be used for various API calls
+func (r *Reconciler) getResourceIDListfromInputSpec(ctx *reconcileRequestContext) error {
+
+	resourceIDListfromSpec := ctx.HostingDeploymentAutoscalingJob.Spec.ResourceID
+	for _, resourceIDfromSpec := range resourceIDListfromSpec {
+		ResourceID := sdkutil.ConvertAutoscalingResourceToString(*resourceIDfromSpec)
```

Comment:
Can this function call result into error/exception ?

Reply:
I though about it, it shouldn't and if it does, it should fail immediately afterwards. Let me know if you insist on having more error handling here.